### DISCUSSION TITLE
pb_hook example to update user record when a related collection is updated?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3339] [Q&A] [answered]
I have a use case for https://pockethost.io where I'd like to update `users.secondsThisMonth` every time `instances.secondsThisMonth` changes. Something like:

```sql
UPDATE users
SET    secondsThisMonth = Sum(instances.secondsThisMonth)
FROM   users
       JOIN instances
         ON users.id = instances.uid
WHERE  users.id = {updatedInstanceRec.uid} 
```

I'd like to do it in a way that the user record receives realtime updates on this metric.

**Comment by ganigeorgiev** [comment, marked as answer]
There are multiple ways you can approach it but here is one example using the [`onModelAfterUpdate` hook](https://pocketbase.io/docs/js-event-hooks/#onmodelafterupdate):

```js
onModelAfterUpdate((e) => {
  // find all instances for the specific user
  const instances = $app.dao().findRecordsByExpr("instances", $dbx.hashExp({ uid: e.model.get("uid" }))

  // sum secondsThisMonth from all instances
  let total = 0;
  for (let instance of instances) {
    total += instances.get("secondsThisMonth");
  }

  // update the instances user with the result
  const user = $app.dao().findRecordById("users", e.model.get("uid"))
  user.set("secondsThisMonth", total)
  $app.dao().saveRecord(user);
}, "instances")
```

_Note that technically you can execute raw queries directly using [`$app.dao().newQuery(...).execute()`](https://pocketbase.io/docs/js-database/#execute) but updating the user record this way will not trigger the users realtime event (while `$app.dao().saveRecord()` on the other hand will trigger it)._

**Reply by ganigeorgiev** [reply]
Another possible solution if you are ok with interval pooling and don't need realtime subscription, could be to create a helper [view collection](https://pocketbase.io/docs/collections/#view-collection) that will compute it for you:
```sql
SELECT
  instances.uid as id,
  instances.uid as user, -- optionally register the field again so that it can be accessed as a relation
  SUM(instances.secondsThisMonth) as totalSecondsThisMonth
FROM instances
GROUP BY instances.uid
```
(_the above should define a new collection with 3 fields - `id`, `user`, and `totalSecondsThisMonth`_)

**Reply by benallfree** [reply]
Is SQLite efficient with caching/indexing views like this?

**Reply by ganigeorgiev** [reply]
SQLite views are **not** persistent/materialized and they technically run on each access (you can think of them as a stored query).

_Another possible solution if you want the value to be computed/dynamic, depending whether the subscribed client have list access to their `instances` records, could be to subscribe to the `instances` and on change retrieve all instances with the JS SDK `getFullList()` and manually sum the field value._

**Reply by benallfree** [reply]
In that case, I think a hook might be better.

Every time the `pocketbase` executable is launched, it creates an `invocations` record that stays updated with total seconds of run time for that invocation. This, in turn, updates the `instances` record with the sum of all invocations from this month. I resorted to a separate SQLite read-only connection to run the `select sum(invocations.totalSeconds)...` query to update the `instances.totalSecondsThisMonth` field. 

**I could maybe move this chunk to a hook, replacing `date-fns` with normal `Date` math**

```ts
    async (instanceId: InstanceId, forPeriod = new Date()) => {
      const startIso = startOfMonth(forPeriod).toISOString()
      const endIso = endOfMonth(forPeriod).toISOString()
      const query = rawDb('invocations')
        .sum('totalSeconds as t')
        .where('instanceId', instanceId)
        .where('startedAt', '>=', startIso)
        .where('startedAt', '<=', endIso)
      raw(query.toString())
      const res = await query
      const [row] = res
      assertExists(row, `Expected row here`)
      const secondsThisMonth = row.t
      await updateInstance(instanceId, { secondsThisMonth })
    },
```

Then, I could update the user model with `totalSecondsThisMonth` to get realtime updates on that if needed. But I don't have a use case for it right now because I think it would be enough to check it at the beginning of each invocation. It doesn't have to be a perfect hard limit.
